### PR TITLE
Feature bridge for Commands & Distributors (Cloud, Local) 

### LINKED ISSUE #647: Refactor TerraHub CLI code to be directly used in Lambda Deployer

### PR DESCRIPTION
**Pull request by aluchianic**
## Description

Fixes #647, #602

## Type of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a change to the documentation

## How Has This Been Tested?

- [ ] to use `--cloud`: 
  - enable `cloudAccount` & `backendAccount` in `.terrahub.yml`
  - set `logs` to `true` in `.terrahub.json`

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have checked that my changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/abstract-command.js [hunk 1]
```diff
@@ -297,6 +303,111 @@ class AbstractCommand {
     this.logger.warn('THUB_TOKEN is not provided.');
     return Promise.resolve();
   }
+
+
+  /**
```
Comment (eistrati) [inline src/abstract-command.js]: remove it pls

File: src/commands/component.js [hunk 2]
```diff
@@ -113,72 +131,72 @@ class ComponentCommand extends AbstractCommand {
...
+      ConfigLoader.writeConfig(existing.config, existing.path);
+    }
 
-        this._createWorkspaceFiles(this._directory);
+    this._createWorkspaceFiles(this._directory);
 
-        const templateName = this._configLoader.getDefaultFileName() + '.twig';
-        const specificConfigPath = path.join(path.dirname(templates.config), templateName);
-        const data = fse.existsSync(specificConfigPath) ? fse.readFileSync(specificConfigPath) : '';
+    const templateName = this._configLoader.getDefaultFileName() + '.twig';
```
Comment (eistrati) [inline src/commands/component.js]: please review the usage of `defaultFileName` across the board and simplify

File: src/abstract-command.js [hunk 3]
```diff
@@ -297,6 +303,111 @@ class AbstractCommand {
...
+        return this.checkProjectDataMissing();
+      });
+    }
+  }
+
+  /**
+   * Check all components dependencies existence
+   * @param {Object} config
+   */
+  checkDependenciesExist(config) {
```
Comment (eistrati) [inline src/abstract-command.js]: check if this function is used in `DistributedCommand` class

File: src/abstract-command.js [hunk 4]
```diff
@@ -297,6 +303,111 @@ class AbstractCommand {
...
+      });
+    }
+
+    return Promise.resolve();
+  }
+
+  /**
+   * @return {Promise}
+   */
+  checkProjectDataMissing() {
```
Comment (eistrati) [inline src/abstract-command.js]: check if this function is used in `DistributedCommand` class

File: src/abstract-command.js [hunk 5]
```diff
@@ -297,6 +303,111 @@ class AbstractCommand {
...
+
+  /**
+   *   ******************* NEW *******************
+   */
+
+  /**
+   * @param {Object} fullConfig
+   * @returns {Promise}
+   */
+  checkProjectDuplicateComponents(fullConfig) {
```
Comment (eistrati) [inline src/abstract-command.js]: check if this function is used in `DistributedCommand` class

File: src/abstract-command.js [hunk 6]
```diff
@@ -297,6 +303,111 @@ class AbstractCommand {
...
+        style: ListException.NUMBER
+      });
+    }
+  }
+
+
+  /**
+   * @returns {Boolean}
+   */
+  isComponentsCountZero() {
```
Comment (eistrati) [inline src/abstract-command.js]: check if this function is used in `DistributedCommand` class

File: src/commands/list.js [hunk 7]
```diff
@@ -50,71 +48,65 @@ class ListCommand extends AbstractCommand {
...
-
-        if (Object.keys(projectsData).length) {
-          this.logger.log('Projects');
-
-          this._showTree(this._format(projectsData, 0, depth));
-        } else {
-          this.logger.log('No Projects to display.');
-        }
+      const sts = new AWS.STS({ credentials });
+      const identity = await sts.getCallerIdentity().promise();
```
Comment (eistrati) [inline src/commands/list.js]: @todo in future, there must be a fall back if user is NOT allowed to call `STS` with action `getCallerIdentity`

File: src/distributed-command.js [hunk 8]
```diff
@@ -47,43 +44,42 @@ class TerraformCommand extends AbstractCommand {
...
       if (nonExistingComponents.length) {
         throw new Error('Some of components were not found: ' + nonExistingComponents.join(', '));
       }
 
-      return Promise.resolve();
-    }).then(() => this._checkProjectDuplicateComponents())
-      .catch(err => {
-        const error = err.constructor === String ? new Error(err) : err;
+      const fullConfig = this.getExtendedConfig();
+      return this.checkProjectDuplicateComponents(fullConfig);
```
Comment (eistrati) [inline src/distributed-command.js]: (1) move this feature into the functionality when terrahub components are loaded, that's supposed to be the moment when they must be checked for duplicates.
(2) don't check for duplicates every time; instead, use `~/.terrahub/cache/` to validate once and check only when the list of components changed

File: src/distributed-command.js [hunk 9]
```diff
@@ -47,43 +44,42 @@ class TerraformCommand extends AbstractCommand {
...
 
-      if (this._tokenIsValid && logs) {
+      if (this._tokenIsValid && this.terrahubCfg.logs) {
         this.logger.updateContext({ canLogBeSentToApi: true });
       }
 
-      return Promise.resolve();
-    }).then(() => this._checkProjectDataMissing()).then(() => {
-      if (this._isComponentsCountZero() && this.getName() !== 'configure') {
+      await this.checkProjectDataMissing();
```
Comment (eistrati) [inline src/distributed-command.js]: (1) move this feature into the functionality when terrahub components are loaded, that's supposed to be the moment when they must be checked for duplicates.
(2) don't check for duplicates every time; instead, use `~/.terrahub/cache/` to validate once and check only when the list of components changed (please research)

File: src/helpers/api-helper.js [hunk 10]
```diff
@@ -378,6 +412,10 @@ class ApiHelper extends events.EventEmitter {
     return this._tokenIsValid;
   }
 
+  get isDeployCloud() {
```
Comment (eistrati) [inline src/helpers/api-helper.js]: please change from `isDeployCloud` to `isCloudDeployer`

File: src/abstract-command.js (hunk 3 above)
Comment (aluchianic) [reply src/abstract-command.js]: transfered to DistributedCommand

File: src/abstract-command.js (hunk 5 above)
Comment (aluchianic) [reply src/abstract-command.js]: transferred to DistributedCommand

File: src/abstract-command.js (hunk 6 above)
Comment (aluchianic) [reply src/abstract-command.js]: transferred to `DistributedCommand`

File: src/abstract-command.js (hunk 4 above)
Comment (aluchianic) [reply src/abstract-command.js]: transferred to `DistributedCommand `

File: src/abstract-command.js (hunk 1 above)
Comment (aluchianic) [reply src/abstract-command.js]: removed

**Comment by eistrati** [comment]
long overdue ... closing this PR and expecting new one instead ...

**Pull request closed without merging**